src/gt4py/next/common.py

```diff
@@ -117,8 +127,81 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:
             else:
                 raise IndexError("UnitRange index out of range")
 
+    def __and__(self, other: UnitRange) -> UnitRange:
+        start = max(self.start, other.start)
+        stop = min(self.stop, other.stop)
+        return UnitRange(start, stop)
+
 
-DomainT: TypeAlias = tuple[tuple[Dimension, UnitRange], ...]
+NamedRange: TypeAlias = tuple[Dimension, UnitRange]
+
+
+@dataclasses.dataclass(frozen=True)
+class Domain(Sequence[NamedRange]):
+    dims: list[Dimension]
+    ranges: list[UnitRange]
+
+    def __post_init__(self):
+        if len(set(self.dims)) != len(self.dims):
+            raise ValueError(f"Domain dimensions must be unique, not {self.dims}.")
+
+    def __len__(self) -> int:
+        return len(self.ranges)
+
+    def __iter__(self):
+        return iter(zip(self.dims, self.ranges))
+
+    @overload
+    def __getitem__(self, index: int) -> NamedRange:
+        ...
+
+    @overload
+    def __getitem__(self, index: slice) -> "Domain":
+        ...
+
+    @overload
+    def __getitem__(self, index: Dimension) -> NamedRange:
+        ...
+
+    def __getitem__(self, index):
+        if isinstance(index, int):
+            return self.dims[index], self.ranges[index]
+        elif isinstance(index, slice):
+            dims_slice = self.dims[index]
+            ranges_slice = self.ranges[index]
+            return Domain(dims_slice, ranges_slice)
+        elif isinstance(index, Dimension):
+            try:
+                index_pos = self.dims.index(index)
+                return self.dims[index_pos], self.ranges[index_pos]
+            except ValueError:
+                raise KeyError(f"No Dimension of type {index} is present in the Domain.")
+        else:
+            raise KeyError("Invalid index type, must be either int, slice, or Dimension.")
+
+    def __and__(self, other: "Domain") -> "Domain":
+        broadcast_dims = promote_dims(self.dims, other.dims)
+        intersected_ranges = [
+            rng1 & rng2
+            for rng1, rng2 in zip(
+                self._broadcast(broadcast_dims, self.dims, self.ranges),
+                self._broadcast(broadcast_dims, other.dims, other.ranges),
+            )
+        ]
+        return Domain(broadcast_dims, intersected_ranges)
+
+    @staticmethod
```

Comment:
I've added this back to Domain as its just a utility method, not really relevant anywhere else.

Reply:
this is against our style guide https://google.github.io/styleguide/pyguide.html#2174-decision